ci-operator/step-registry/mco/conf/day2/create-custom-mcp/mco-conf-day2-create-custom-mcp-commands.sh

```diff
@@ -60,32 +73,46 @@ EOF
     echo "It has been requested to add 0 nodes to the pool. No node will be added to the custom MachineConfigPool."
     return
   fi
-  [ -z "$MCP_NUM_NODES" ] \
-    && echo "MCO_CONF_DAY2_CUSTOM_MCP_NUM_NODES variable is empty. All nodes matching the '$MCP_LABEL' label will be added to the custom pool" \
-    || echo "Labeling $MCP_NUM_NODES worker nodes in order to add them to the new custom MCP"
-  # shellcheck disable=SC2046
-  oc label node $(oc get nodes -l "$MCP_LABEL" -ojsonpath="{.items[:$MCP_NUM_NODES].metadata.name}") node-role.kubernetes.io/"$MCO_CONF_DAY2_CUSTOM_MCP_NAME"=
+  if (( ${#NODES_LIST[@]} == 0 )); then
```

Comment:
suggest to add a checkpoint here to make sure node num in NODES_LIST >= MCP_NUM_NODES

Reply:
So the node num is not a sensitive param, by default we will create multiple pools with same size if possible
Is there any requirement to create different pools with different size, e.g. I have a mix-arch cluster, the cluster has 3 workers, 1 amd64 worker, 2 arm64 workers, then I want to create 2 pools for different arches separately.

Reply:
Actually I have no specific requirement to create different pools with different size, but during my coding, I think that is also a possible case. E.g: 3 workers in total, but set `MCP_NUM_NODES` to 2, then the 2nd mcp's size will be different with the 1st one. Of course, if users want to create multiple pools with same size, they need to adjust the setting of `MCP_NUM_NODES` or scale up worker number to make the two mcp size the same. But from code perspective, that is NOT a MUST requirement.

Reply:
`same size` here means we only support user to configure the pool size for all the pools by param `MCO_CONF_DAY2_CUSTOM_MCP_NUM_NODES`, to make this step more robust, It's better to create a pool with 2 params, `label` and `size`. e.g. according to above scenario, if I want to create 2 pools for arm64 and amd64 workers, I need to filter out the nodes by arch related `label`,then label the `size` nodes with new mcp name

Reply:
I did not hear it is a MUST that have to two mcp with the same size, I think that totally depends on user's desire. 

> according to above scenario

Are you talking about "I have a mix-arch cluster, the cluster has 3 workers, 1 amd64 worker, 2 arm64 workers, then I want to create 2 pools for different arches separately" ? What size you are want to set for each mcp ?

> if I want to create 2 pools for arm64 and amd64 workers, I need to filter out the nodes by arch related label,then label the size nodes with new mcp name

I think that is reasonable, what behavior are you expecting ?

Reply:
> Are you talking about "I have a mix-arch cluster, the cluster has 3 workers, 1 amd64 worker, 2 arm64 workers, then I want to create 2 pools for different arches separately" ? What size you are want to set for each mcp ?

Yes

> I think that is reasonable, what behavior are you expecting ?

for example, I have a mix-arch cluster provision step, I can configure node num for different arches, so before creating custom pool I already know the pool size of different arches

then we can use 3 params to configure pool name, pool size and node label e.g.
`MCO_CONF_DAY2_CUSTOM_MCP_NAME=mcp4amd64 mcp4arm64`
`MCO_CONF_DAY2_CUSTOM_MCP_FROM_LABEL=kubernetes.io/arch=amd64 kubernetes.io/arch=arm64`
`MCO_CONF_DAY2_CUSTOM_MCP_NUM_NODES=1 2`

for every pool we have 3 params, the data struct is like below 
```
{
   "pool_name": "mcp4amd64",
   "pool_size": 1,
   "node_label": "kubernetes.io/arch=amd64"
}
```

To create every pool the logic is like, fill the `NODES_LIST` by `node_label`, create the custom pool with `pool_name`, loop the `NODES_LIST` to get nodes by `pool_size` and label these nodes with `pool_name`

Reply:
@sergiordlr @aleskandro WDYT

Reply:
> MCO_CONF_DAY2_CUSTOM_MCP_NAME=mcp4amd64 mcp4arm64
> MCO_CONF_DAY2_CUSTOM_MCP_FROM_LABEL=kubernetes.io/arch=amd64 kubernetes.io/arch=arm64
> MCO_CONF_DAY2_CUSTOM_MCP_NUM_NODES=1 2

Do you mean you want `MCO_CONF_DAY2_CUSTOM_MCP_FROM_LABEL` and `MCO_CONF_DAY2_CUSTOM_MCP_NUM_NODES` to support multi values, even they are mapped in order?

```
{ "pool_name": "mcp4amd64", "pool_size": 1, "node_label": "kubernetes.io/arch=amd64" }
{ "pool_name": "mcp4arm64", "pool_size": 2, "node_label": "kubernetes.io/arch=arm64" }
```

If my understand is right, the parameter should be kind of hash dict data, but not use 3 ENV var to map them in order, it is not a robust way. If we want the accurate mcp definition, we need to a big change to replace the 3 ENV with a new ENV using json format value.  

In current implementation, it is supposed to support multi mcp by only setting `MCO_CONF_DAY2_CUSTOM_MCP_NAME`, both `MCO_CONF_DAY2_CUSTOM_MCP_FROM_LABEL` and `MCO_CONF_DAY2_CUSTOM_MCP_NUM_NODES` are shared for each mcp creation.

Reply:
Just like what I explained in the above, the current implemention do NOT support that, it only support create multi mcp with the same label filter and size setting. I do not have plan to change it too much to avoid impact on the existing ci jobs.

Reply:
I like @rioliu-rh idea of defining a better 'data model' for this step since ever. However, I also understand the complexity of that. 

Can you clarify what are the reasons triggering the changes in this step now? Is there a new feature that is going to make use of this step?

Reply:
Right, this change is from a test requirement from a slack [discussion](https://redhat-internal.slack.com/archives/GH9JHQFKN/p1721311802315329?thread_ts=1721107446.852209&cid=GH9JHQFKN).

If everyone like the idea of 'data model', I can have a shot.